File: Medium/1. Shopping Cart Problem/solutions/AKSHIT989_Akshit/medium.py
from prettytable import PrettyTable
#shopping items
shoppingItems = {
   #no: [     name             ,price, discounted_price,  weight      ]
    1 : ['Basshead Earphones'  ,1200 , 899   , 150] ,
    2 : ['Bluetooth Mouse'     ,600  , 600   , 120] ,
    3 : ['32-GB Pen drive'     ,499  , 399   , 30] ,
    4 : ['29-Inch LED monitor' ,33499, 29999 , 4000] ,
    5 : ['i7-processor'        ,18999, 18999 , 15] ,
    6 : ['1-TB Hard Drive'     ,3999 , 3999  , 175] ,
    7 : ['16-GB DDR4 RAM'      ,10999, 9999  , 35]
}
y=PrettyTable()
y.field_names = ["Item No.", "Item Name", "Price","Discount price","Weight"]
for i in shoppingItems:
    y.add_row([i , shoppingItems[i][0], shoppingItems[i][1], shoppingItems[i][2], shoppingItems[i][3] ])

global total_bill
total_bill=0
discounted_bill=0

#customer data input
Name      = input('\nEnter your Name: ')
phone     = input('Enter Contact No.: ')
payMethod = input('Enter payment method(cash/card/online): ')

bill = { } #storing purchased items

#shopping list input starts from here
while(True):
    print('\nShopping list::\n')
    print(y)
    
    choice=int(input('Enter your choice:'))
    quantity=int(input('Enter quantity:'))
    bill[choice]=quantity
    total_bill += shoppingItems[choice][1]*quantity
    discounted_bill+=shoppingItems[choice][2]*quantity
    opt=input('\nDo you want to continue:(Y-Yes/N-No):')
    if opt=='n':
        break
#shopping list input ends here
dist=0
address=''

thd=input('Select T-Takeaway/H-Home Delivery: ')
if thd=='h' :
    dist=int(input('Distance from shop to the delivery address(in km): '))
    address=input('Shipping Address: ')

ship_charge=0
if dist<=5:
    ship_charge=0
elif dist<=20:
    ship_charge=30
elif dist<=50:
    ship_charge=60
else:
    ship_charge=-1


#BILL GENERATION
print('-------------------------------------------------------------------')
print('                           Your bill:                            \n')
#printing static data
print(
    '                                        Shop name: GadgetifyWithGSBlr\n',
    '                                    Shop address: 311/5 Akshay nagar,\n',
    '                                          Bangalore, Karnataka, India\n'
    '                                      Shop contact no: +91 9988776655\n'
)
#end of printing static data
#printing variable data
print('\nCustomer name: '+ Name)
print('Contact No.: '+ phone)
print('\nItems bought:')


x = PrettyTable()
x.field_names = ["Item Name", "Quantity", "Price","Discount price"]

for i in bill:
    x.add_row([shoppingItems[i][0],bill[i] , shoppingItems[i][1]*bill[i], shoppingItems[i][2]*bill[i] ])
print(x)

print('\nTotal tax on original price: %.2f' %(total_bill*0.06))
print('\nTotal tax on discount price: %.2f' %(discounted_bill*0.06))

if ship_charge==-1:
    print('Shipping not possible')
    ship_charge=0
else:
    print('Total Shipping charges: ',ship_charge)

print('Total Amount saved: ', total_bill-discounted_bill)

print('Amount to be paid: %.2f' %(discounted_bill*1.06 + ship_charge))
print('\nPayment method used:'+ payMethod)

import datetime
now = datetime.datetime.now()
print ('Date and time:',now.strftime("%d-%m-%Y %H:%M:%S"))

if thd=='h':
    print('Shipping Address:', address)

print('-------------------------------------------------------------------')

File: Medium/1. Shopping Cart Problem/solutions/nikhil7sunny_NIKHIL/medium.cpp
#include<bits/stdc++.h>
#include<string>
using namespace std;

struct user{
	string uname;
	string phone ;
	string paymethod;
	string it[10];
	int quantity[10];
	string deliveryType;
	string address;
	float distance;
	float deliveryfee;
	float net[10];
	float total;
	int n;
}user_1;

struct shop{
	string sname ;
	string saddress;
	string sphone;
}shop_1;

struct item{
	string iname;
	float weight;
	float price;
	float discountPrice;

}item_list[7];

int size = 7,cnt = 0;
float tax = 0.06;
time_t now = time(0);
char* dt =ctime(&now);

void loop(){
	cout<<"\n";
	for (int i = 1;i<74;i++)
		cout<<"*";
}

void listassign(){
	item_list[0].iname = "Redmi Note 9 Pro Max";
	item_list[0].price = 16499;
	item_list[0].weight = 209;
	item_list[0].discountPrice = 15999;
	item_list[1].iname = "Redmi Note 9 Pro";
	item_list[1].price = 12999;
	item_list[1].weight = 210;
	item_list[1].discountPrice = 12000;
	item_list[2].iname = "Redmi 8A Dual";
	item_list[2].price = 6499;
	item_list[2].weight = 212.66;
	item_list[2].discountPrice = 0;
	item_list[3].iname = "Redmi 10 Pro";
	item_list[3].price = 50499;
	item_list[3].weight = 211;
	item_list[3].discountPrice = 45000;
	item_list[4].iname = "Redmi 10 Max";
	item_list[4].price = 49000;
	item_list[4].weight = 211;
	item_list[4].discountPrice = 43000;
	item_list[5].iname = "Redmi K30";
	item_list[5].price = 16100;
	item_list[5].weight = 230;
	item_list[5].discountPrice = 15400;
	item_list[6].iname = "Redmi A3";
	item_list[6].price = 12999;
	item_list[6].weight = 215;
	item_list[6].discountPrice = 12000;
	shop_1.sname = "GadgetifyWithGSBlr";
	shop_1.saddress = "311/5 Akshay nagar, Bangalore, Karnataka, India";
	shop_1.sphone = "+91 9988776655";
}

void displaylist(){
	cout<<setw(15)<<"ITEM NAME"<<setw(18)<<"PRICE"<<setw(19)<<"DISCOUNT PRICE"<<setw(10)<<"WEIGHT"<<"\n";
	cout<<setw(15)<<"---------"<<setw(18)<<"-----"<<setw(19)<<"--------------"<<setw(10)<<"------"<<"\n";
	for (int i = 0 ;i<size ;i++){
		cout<<setw(21)<<item_list[i].iname<<setw(8)<<"Rs "<<item_list[i].price;
		if(item_list[i].discountPrice != 0)
			cout<<setw(15)<<item_list[i].discountPrice;
		else
			cout<<setw(19)<<"No Discount";
		cout<<setw(10)<<item_list[i].weight<<"\n";

	}
}

void user_details(){
	string choice;
	cout<<"\n Enter the User Details :\n\n";
	cout<<setw(40)<<"Name : ";
	getline(cin,user_1.uname);
	cout<<setw(40)<<"Phone No : ";
	getline(cin,user_1.phone);
	cout<<setw(40)<<"Payment Method (cash/card/online) : ";
	getline(cin,user_1.paymethod);
	do{	
		cout<<setw(40)<<"Item to be Purchased : ";
		getline(cin,user_1.it[cnt]);
		cout<<setw(40)<<"Enter quantity : ";
		cin>>user_1.quantity[cnt];
		cin.ignore (std::numeric_limits<std::streamsize>::max(), '\n');
		cout<<setw(40)<<"Do You want to add another item to the cart\n";
		getline(cin,choice);
		cnt++;
	}while(choice.compare("yes") == 0);
	cout<<"\n Items successfully added to Cart\n";
	cout<<setw(40)<<"Delivery Type (Home Delivery/Take Away) : ";
	getline(cin,user_1.deliveryType);
	if(user_1.deliveryType.compare("Home Delivery") == 0){
		cout<<setw(40)<<"Distance : ";
		cin>>user_1.distance;
		cin.ignore (std::numeric_limits<std::streamsize>::max(), '\n');
		if(user_1.distance > 50)
			cout<<setw(40)<<"Cannot Deliver at a distance greater than 50 !!!!";
		else{
			cout<<setw(40)<<"Address :";
			getline(cin,user_1.address);
		}
	}
	cout<<"\n Your Order has been successfully Placed";
}

void caldelivery(){
	if(user_1.distance < 50 && user_1.distance > 20)
		user_1.deliveryfee = 60;
	else if (user_1.distance >20 && user_1.distance > 5)
		user_1.deliveryfee = 30;
	else
		user_1.deliveryfee = 0;
}

float netcalculate(){
	float sum = 0;
	user_1.total = 0;
	for (int i=0;i<cnt;i++){
		for(int j=0;j<size;j++){
			if(user_1.it[i].compare(item_list[j].iname) == 0){
				if(item_list[j].discountPrice == 0)
					user_1.net[i] = item_list[j].price * user_1.quantity[i];
				else 
					user_1.net[i] = item_list[j].discountPrice * user_1.quantity[i];
				user_1.total += item_list[j].price * user_1.quantity[i];
				break;
			}
		}
	sum += user_1.net[i];	
	}
	return sum;
}

void generatebill(){
	float tprice = netcalculate();
	int flag = 0;
	cout<<"\n\n\n";
	cout<<"\t\t"<<dt<<"\n"<<"\t\t\t"<<shop_1.sname;
	cout<<"\n"<<"\t\t\t------------------\n"<<setw(15)<<shop_1.sphone;
	// cout<<"\n"<<shop_1.saddress;
	cout<<setw(26);
	for(int i = 0;i<shop_1.saddress.size();i++){
		cout<<shop_1.saddress[i];
		if(shop_1.saddress[i] == ','){
			cout<<"\n"<<setw(40);
		}
	}
	cout<<"\n\n";
	cout<<"\t\t\t BILL GENERATED\n";
	cout<<"\t\t\t --------------";
	cout<<"\n"<<setw(15)<<"User Details";
	cout<<"\n"<<setw(30)<<"Name : "<<user_1.uname;
	cout<<"\n"<<setw(30)<<"Phone NO : "<<user_1.phone;
	cout<<"\n"<<setw(30)<<"Payment Method : "<<user_1.paymethod;
	cout<<"\n"<<setw(30)<<"Delivery Type : "<<user_1.deliveryType;
	if(user_1.deliveryType.compare("Home Delivery") == 0 && user_1.distance <50){
		cout<<"\n"<<setw(30)<<"Distance : "<<user_1.distance;
		cout<<"\n"<<setw(30)<<"Address : "<<user_1.address;
		flag = 1;
	}
	cout<<"\n\n"<<setw(15)<<"ITEM"<<setw(18)<<"QUANTITY"<<setw(15)<<"PRICE";
	cout<<"\n"<<setw(15)<<"----"<<setw(18)<<"--------"<<setw(15)<<"-----";
	for (int i=0;i<cnt;i++)
		cout<<"\n"<<setw(13)<<user_1.it[i]<<setw(18)<<user_1.quantity[i]<<setw(15)<<user_1.net[i];
	cout<<"\n"<<setw(50)<<"Amount Saved :  Rs "<<user_1.total-tprice;
	cout<<"\n"<<setw(50)<<"tax : "<<tax*100<<" %";
	tprice = tprice*(1+tax);
	if(flag){
		cout<<"\n"<<setw(50)<<"Delivery Fee : "<<user_1.deliveryfee;
		tprice = tprice + user_1.deliveryfee;
	}
	cout<<"\n"<<setw(50)<<"Total To Be Payed : "<<tprice;
}

int main(){
	listassign();
	displaylist();
	user_details();
	caldelivery();
	loop();
	generatebill();
	loop();
}